File: src/diffvouch/render.py
from __future__ import annotations

import json

from diffvouch.models import Finding, ReviewResult


def render_json(result: ReviewResult) -> str:
    return (
        json.dumps(result.model_dump(mode="json", by_alias=True), indent=2, sort_keys=True) + "\n"
    )


def _finding(finding: Finding) -> list[str]:
    location = ""
    if finding.path:
        location = finding.path
        if finding.line:
            location += f":{finding.line}"
        location = f" — {location}"
    kind = "BLOCKING" if finding.blocking else "NON-BLOCKING"
    return [
        f"[{kind}] [{finding.severity.value.upper()}] {finding.id}: {finding.title}{location}",
        f"  {finding.explanation}",
        f"  Recommendation: {finding.recommendation}",
    ]


def render_terminal(result: ReviewResult) -> str:
    lines = [
        "DiffVouch Review",
        "=" * 16,
        f"Rating: {result.rating.overall:.1f}/5 — {result.rating.label}",
        f"Scope: {result.scope.mode} ({result.scope.base_ref})",
        (
            f"Provider: {result.provider.name}/{result.provider.transport} "
            f"model={result.provider.model}"
        ),
        f"Summary: {result.summary}",
        "",
        "Rating breakdown",
    ]
    for name, score in result.rating.dimensions.model_dump().items():
        lines.append(f"  {name:<16} {score:.1f}/5")

    blocking = [finding for finding in result.findings if finding.blocking]
    non_blocking = [finding for finding in result.findings if not finding.blocking]
    for title, findings in (("Blocking issues", blocking), ("Non-blocking issues", non_blocking)):
        lines.extend(["", title])
        if not findings:
            lines.append("  None")
        for finding in findings:
            lines.extend(_finding(finding))

    if result.needs_verification:
        lines.extend(["", "Needs verification"])
        lines.extend(f"  - {item}" for item in result.needs_verification)
    if result.positive_observations:
        lines.extend(["", "Positive observations"])
        lines.extend(f"  - {item}" for item in result.positive_observations)

    lines.extend(
        [
            "",
            f"Files reviewed: {len(result.files.reviewed)}",
            f"Gate: {'passed' if result.gate.passed else 'failed'}",
        ]
    )
    if result.gate.reasons:
        lines.extend(f"  - {reason}" for reason in result.gate.reasons)
    if result.publication.published:
        lines.append(f"Published by {result.publication.bot}: {result.publication.url}")
    elif result.publication.requested:
        lines.append("GitHub publication requested but not completed.")
    else:
        lines.append("Not published (use --publish to post a GitHub App review).")
    return "\n".join(lines) + "\n"

File: internal/render/render.go
package render

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/divyangchauhan/DiffVouch/internal/gitdiff"
	"github.com/divyangchauhan/DiffVouch/internal/model"
)

func JSON(result model.ReviewResult) (string, error) {
	raw, err := json.MarshalIndent(result, "", "  ")
	if err != nil {
		return "", err
	}
	return string(raw) + "\n", nil
}

func Terminal(result model.ReviewResult) string {
	var output strings.Builder
	fmt.Fprintf(&output, "DiffVouch review: %.1f/5 — %s\n\n%s\n\n", result.Rating.Overall, result.Rating.Label, result.Summary)
	fmt.Fprintf(&output, "Scope: %s · base %s (%s) · head %s\n\n", result.Scope.Mode, result.Scope.BaseRef, result.Scope.BaseSHA, result.Scope.HeadSHA)
	blocking := filter(result.Findings, true)
	nonBlocking := filter(result.Findings, false)
	writeFindings(&output, "Blocking issues", blocking)
	writeFindings(&output, "Non-blocking issues", nonBlocking)
	if len(result.PositiveObservations) > 0 {
		output.WriteString("Positive observations\n")
		for _, observation := range result.PositiveObservations {
			fmt.Fprintf(&output, "  - %s\n", observation)
		}
		output.WriteString("\n")
	}
	if len(result.NeedsVerification) > 0 {
		output.WriteString("Needs verification\n")
		for _, item := range result.NeedsVerification {
			fmt.Fprintf(&output, "  - %s\n", item)
		}
		output.WriteString("\n")
	}
	fmt.Fprintf(&output, "Rating breakdown\n  Correctness: %.1f  Security: %.1f  Maintainability: %.1f  Testing: %.1f  Scope: %.1f\n\n",
		result.Rating.Dimensions.Correctness, result.Rating.Dimensions.Security,
		result.Rating.Dimensions.Maintainability, result.Rating.Dimensions.Testing,
		result.Rating.Dimensions.Scope)
	fmt.Fprintf(&output, "Reviewed %d files with %s/%s", len(result.Files.Reviewed), result.Provider.Name, result.Provider.Transport)
	if result.Provider.Model != "" {
		fmt.Fprintf(&output, " using %s", result.Provider.Model)
	}
	output.WriteString(".\n")
	if result.Files.Redactions > 0 {
		fmt.Fprintf(&output, "%d potential secret(s) were redacted before provider submission.\n", result.Files.Redactions)
	}
	if len(result.Files.Binary) > 0 {
		fmt.Fprintf(&output, "Skipped binary or non-UTF-8 files: %s\n", strings.Join(displayPaths(result.Files.Binary), ", "))
	}
	if len(result.Files.Excluded) > 0 {
		fmt.Fprintf(&output, "Excluded files: %s\n", strings.Join(displayPaths(result.Files.Excluded), ", "))
	}
	if !result.Gate.Passed {
		fmt.Fprintf(&output, "Quality gate failed: %s\n", strings.Join(result.Gate.Reasons, "; "))
	} else {
		output.WriteString("Quality gate passed.\n")
	}
	if result.Publication.Published && result.Publication.URL != nil {
		fmt.Fprintf(&output, "Published by %s: %s\n", pointer(result.Publication.Bot), *result.Publication.URL)
	} else if !result.Publication.Requested {
		output.WriteString("Nothing was published.\n")
	}
	return output.String()
}

func filter(findings []model.Finding, blocking bool) []model.Finding {
	var result []model.Finding
	for _, finding := range findings {
		if finding.Blocking == blocking {
			result = append(result, finding)
		}
	}
	return result
}

func writeFindings(output *strings.Builder, heading string, findings []model.Finding) {
	output.WriteString(heading + "\n")
	if len(findings) == 0 {
		output.WriteString("  None.\n\n")
		return
	}
	for _, finding := range findings {
		location := ""
		if finding.Path != nil {
			location = " " + gitdiff.DisplayPath(*finding.Path)
			if finding.Line != nil {
				location += fmt.Sprintf(":%d", *finding.Line)
			}
		}
		fmt.Fprintf(output, "  %s [%s] %s%s\n    %s\n    Recommendation: %s\n", finding.ID, finding.Severity, finding.Title, location, finding.Explanation, finding.Recommendation)
	}
	output.WriteString("\n")
}

func displayPaths(paths []string) []string {
	display := make([]string, len(paths))
	for index, path := range paths {
		display[index] = gitdiff.DisplayPath(path)
	}
	return display
}

func pointer(value *string) string {
	if value == nil {
		return "configured bot"
	}
	return *value
}
